File: src/list.ts
let list32_create: (inital_allocated_length: number) => number;
let list64_create: (inital_allocated_length: number) => number;
let list32_push: (ptr: number, value: number) => null;
let list64_push: (ptr: number, value: number) => null;
let list_free: (ptr: number) => null;

const listRegistry = new FinalizationRegistry((ptr: number) => {
    list_free(ptr);
});

export function init() {
    list32_create = cwrap("list32_create", "number", ["number"]);
    list64_create = cwrap("list64_create", "number", ["number"]);
    list32_push = cwrap("list_push32", null, ["number", "number"]);
    list64_push = cwrap("list_push64", null, ["number", "number"]);
    list_free = cwrap("list_free", null, ["number"]);
}


export class List {
    readonly ptr: number;
    readonly elementSize: 4 | 8;

    static create32(initalAllocatedLength: number) {
        const ptr = list32_create(initalAllocatedLength);
        const list = new List(ptr, 4)
        listRegistry.register(list, list.ptr);
    }

    static create64(initalAllocatedLength: number) {
        const ptr = list64_create(initalAllocatedLength);
        const list = new List(ptr, 8)
        listRegistry.register(list, list.ptr);
    }

    static adopt32(ptr: number) {
        const list = new List(ptr, 4);
        listRegistry.register(list, list.ptr);
        return list;
    }

    static adopt64(ptr: number) {
        const list = new List(ptr, 8);
        listRegistry.register(list, list.ptr);
        return list;
    }

    constructor(ptr: number, elementSize: 4 | 8) {
        this.ptr = ptr;
        this.elementSize = elementSize;
    }

    get length() {
        return getValue(this.ptr + 0, "i32");
    }

    get allocatedLength() {
        return getValue(this.ptr + 4, "i32");
    }

    at(index: number) {
        index = Math.trunc(index);
        if (index < 0) index = this.length - index;
        if (index >= this.length) throw RangeError("Out of bounds list access");

        const elementsPtr = getValue(this.ptr + 8, "i32*");
        if (this.elementSize == 4) {
            return getValue(elementsPtr + index * 4, "i32");
        } else {
            let lower = getValue(elementsPtr + index * 8, "i32");
            let upper = getValue(elementsPtr + index * 8 + 4, "i32");
            if (upper < 0)
                lower = ~lower + 1;

            return upper * (2 ** 32) + lower;
        }
    }

    push(value: number) {
        if (this.elementSize == 4) {
            list32_push(this.ptr, value);
        } else {
            list64_push(this.ptr, value);
        }
    }

    free() {
        list_free(this.ptr);
    }
}

File: src/list.c
#include "list.h"

void list_free(ListAbstract* list) {
	free_inst(list->elements);
	free_inst(list);
}

void list_allocate_length(ListAbstract* list, int min_length, int element_size) {
	if (list->allocated_length >= min_length) return;

	list->allocated_length *= 2;
	list->elements = realloc_inst(list->elements, list->allocated_length * element_size);

	if (list == NULL) {
		fprintf(stderr, "Failed to allocate memory: list_allocate_length()\n");
		exit(1);
	}
}

void list32_push(List32* list, uint32_t value) {
	list_allocate_length(list, list->length + 1, 4);

	*(uint32_t*)(list->elements + list->length * 4) = value;
	list->length++;
}

void list64_push(List64* list, uint64_t value) {
	list_allocate_length(list, list->length + 1, 8);

	*(uint64_t*)(list->elements + list->length * 8) = value;
	list->length++;
}

uint32_t list32_at(List32* list, int index) {
	return *(uint32_t*)(list->elements + index * 4);
}

uint64_t list64_at(List64* list, int index) {
	return *(uint64_t*)(list->elements + index * 8);
}

ListAbstract* list_create(int element_size, int inital_allocated_length) {
	ListAbstract* list = malloc_inst(sizeof(ListAbstract));

	if (list == NULL) {
		fprintf(stderr, "Failed to allocate memory: list_create()\n");
		exit(1);
	}

	list->elements = malloc_inst(inital_allocated_length * element_size);

	if (list->elements == NULL) {
		fprintf(stderr, "Failed to allocate memory: list_create()\n");
		exit(1);
	}

	list->length = 0;
	list->allocated_length = inital_allocated_length;

	return list;
}

List32* list32_create(int inital_allocated_length) {
	return (List32*)list_create(4, inital_allocated_length);
}

List64* list64_create(int inital_allocated_length) {
	return (List64*)list_create(4, inital_allocated_length);
}